pagespeed/automatic/proxy_fetch.cc

```diff
@@ -90,6 +91,37 @@ ProxyFetchFactory::~ProxyFetchFactory() {
             << " outstanding requests.";
 }
 
+void ProxyFetchFactory::CancelOutstanding() {
+  int64 sleep_us = 250;
+
+  // First wait any current scheduled CompleteFinishParse calls to round up,
+  // so we don't have to worry about races w/regard to done_outstanding_ and
+  // finishing_, avoiding the need to take a lock on proxy_fetch.
+  while (proxy_fetches_done_in_flight_.value() != 0) {
+    timer_->SleepUs(sleep_us);
+  }
+
+  // Any outstanding fetches left do not have a Done() call initiated on them.
+  // So we don't have to worry about done_outstanding_ and finishing_
+  while (true) {
+    ProxyFetch* fetch;
+    {
+      ScopedMutex lock(outstanding_proxy_fetches_mutex_.get());
+      if (outstanding_proxy_fetches_.empty()) {
+        break;
+      }
+      fetch = *outstanding_proxy_fetches_.begin();
+      outstanding_proxy_fetches_.erase(fetch);
+    }
+    fetch->Done(false);
```

Comment:
I don't really see anything that's preventing someone else from calling Done on the fetch... or have we shut down the threads feeding it data at this point?